mdata/cache/cache_metric.go

```diff
@@ -0,0 +1,236 @@
+package cache
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/raintank/metrictank/mdata/chunk"
+)
+
+type CCacheMetric struct {
+	sync.RWMutex
+	oldest uint32
+	newest uint32
+	chunks map[uint32]*CacheChunk
+}
+
+func NewCCacheMetric() *CCacheMetric {
+	return &CCacheMetric{
+		chunks: make(map[uint32]*CacheChunk),
+	}
+}
+
+func (mc *CCacheMetric) Init(prev uint32, itergen chunk.IterGen) bool {
+	res := mc.Add(prev, itergen)
+	if !res {
+		return false
+	}
+	mc.oldest = itergen.Ts()
+	mc.newest = itergen.Ts()
+	return true
+}
+
+func (mc *CCacheMetric) Del(ts uint32) int {
+	mc.Lock()
+	defer mc.Unlock()
+
+	if _, ok := mc.chunks[ts]; !ok {
+		return len(mc.chunks)
+	}
+
+	prev := mc.chunks[ts].Prev
+	next := mc.chunks[ts].Next
+
+	if prev != 0 {
+		mc.chunks[prev].Next = mc.chunks[ts].Next
+	}
+	if next != 0 {
+		mc.chunks[next].Prev = mc.chunks[ts].Prev
+	}
+
+	return len(mc.chunks)
+}
+
+func (mc *CCacheMetric) Add(prev uint32, itergen chunk.IterGen) bool {
+	var ts, endTs uint32
+	ts = itergen.Ts()
+
+	mc.Lock()
+	defer mc.Unlock()
+	if _, ok := mc.chunks[ts]; ok {
+		return false
+	}
+
+	mc.chunks[ts] = &CacheChunk{
+		Ts:    ts,
+		Prev:  prev,
+		Itgen: itergen,
+	}
+
+	// if the previous chunk is cached, set this one as it's next
+	if _, ok := mc.chunks[prev]; ok {
+		mc.chunks[prev].Next = ts
+	}
+
+	endTs = mc.EndTs(ts)
+
+	// if EndTs() can't figure out the end date it returns ts
+	if endTs != ts {
+		if _, ok := mc.chunks[endTs]; ok {
+			mc.chunks[endTs].Prev = ts
+		}
+	}
+
+	// update list head/tail if necessary
+	if ts > mc.newest {
+		mc.newest = ts
+	} else if ts < mc.oldest {
+		mc.oldest = ts
+	}
+
+	return true
+}
+
+// get sorted slice of all chunk timestamps
+// assumes we have at least read lock
+func (mc *CCacheMetric) sortedTs() *[]uint32 {
+	keys := make([]uint32, 0, len(mc.chunks))
+	for k := range mc.chunks {
+		keys = append(keys, k)
+	}
+	sort.Sort(uint32Asc(keys))
+	return &keys
+}
+
+// takes a chunk's ts and returns the end ts (guessing if necessary)
+// assumes we already have at least a read lock
+func (mc *CCacheMetric) EndTs(ts uint32) uint32 {
+	span := (*mc.chunks[ts]).Itgen.Span()
+	if span > 0 {
+		// if the chunk is span-aware we don't need anything else
+		return (*mc.chunks[ts]).Ts + span
+	}
+
+	if (*mc.chunks[ts]).Next == 0 {
+		if (*mc.chunks[ts]).Prev == 0 {
+			// if a chunk has no next and no previous chunk we have to assume it's length is 0
```

Comment:
actually i think it is important to state that we assume length `0`. this has implications like for example that nothing will try to connect a chunk of `length 0` with a following chunk, because we never know if there isn't another in between the two.